src/components/shoppingListItems.js

```diff
@@ -55,14 +45,20 @@ const shoppingListItemSearch = gql`
 `
 
 const Listing = () => {
-  const { data, loading, error } = useQuery(
-    shoppingListItemSearch,
-    {
-      fetchPolicy: 'cache-and-network'
-    }
-  )
+  const { data, loading, error } = useQuery(shoppingListItemSearch, { fetchPolicy: 'cache-and-network' })
+  const toggleStatus = (id, currentStatus) => {
+    console.log(id, currentStatus)
+    const { data, loading, error } = useMutation(
+      shoppingListItemUpdate,
+      { variables: { id: id, status: statusEnum[currentStatus] } }
+    )
+
+    if (loading) { return(LoadingIndicator) }
+    if (error) { return handleError(error) }
+    if (data) return(data.shoppingListItemUpdate)
+  }
```

Comment:
I know Lines 55+ don't work and probably need to make `useQuery` and `useMutation` end by updating some state data. However, I am getting an error because of an invalid hook and I am unsure of the cause. If you have an idea, I'd love to learn, Fynn

Reply:
@dostaglou 

Lets start by your error message:

> Error: Invalid hook call. Hooks can only be called inside of the body of a function component

This error happens because you are calling a hook (i.d. `useXXX(() => {}, []`) inside a function. But hooks can only be called inside a react function component (i.d. an arrow function which returns a JSX element, simply said).

**Normal Function** (returns anything except a JSX element)
() => { return any }

**Function component** (because it returns an element (div, span, custom elements, etc.)
() => { return <Element />

You can find a good read about your specific problem here: [Only Call Hooks from React Functions](https://reactjs.org/docs/hooks-rules.html#only-call-hooks-from-react-functions)


So, how do we fix it?
First, we are going to put the mutation on the same top level as your query.
Secondly, we are going to return a tuple on the useMutation hook which consist of a mutate function and the object you already have with the current status of your fields.

E.g. 
```
const [updateListItem, { data, loading, error }] = useMutation(shoppingListItemUpdate)
```

Now we can use `updateListItem` in your `toggleStatus()` like this `updateListItem({ variables: { id: id, status: statusEnum[currentStatus] } })`


What about handling the returned `loading` and `error` status?
I would rename these variables to make them distinct from the useQuery hook results and based on their value I would show the same loading indicator or a specific error message:
E.g.
```
if (loadingQuery || loadingMutation) return <LoadingIndicator />
```
(BTW your LoadingIndicator is a React function component and can therefore be put into `< />` to make it clearer)


I think this should do the trick for now. 
Just keep in mind that when the state of a hook changes, usually the component re-renders and has therefore a new state which can be utilized. Reacts concept is a little bit different than other libraries but once you understand it, I think it a great library for FE development!

Reply:
I think I may still not be understanding something. 

When attempting to insert the following: 
`const [updateListItem, { data, loading, error }] = useMutation(shoppingListItemUpdate)`
I get an error: `identifier 'data' has already been declared.`

I supposed this came from the fact that right above it I had: 
`  const { data, loading, error } = useQuery(shoppingListItemSearch, { fetchPolicy: 'cache-and-network' })`

So I tried swapping that out for:
` const [updateListItem, mutationData] = useMutation(updateShoppingListItem)`

This no longer had the error ✅ . Then I added the following: 
` const toggleStatus = (id, currentStatus) => { updateListItem({variables: { id: id, status: statusEnum[currentStatus] } }) } `

Thus inside the Listing function, I had the following: 
```
  const { data, loading, error } = useQuery(shoppingListItemSearch, { fetchPolicy: 'cache-and-network' })
  const [updateListItem, mutationData] = useMutation(updateShoppingListItem)

  const toggleStatus = (id, currentStatus) => { updateListItem({variables: { id: id, status: statusEnum[currentStatus] } }) }
  
  // if code
 // return < component >
```

That formulation seemed to call the mutation but it didn't pass the variables ( my server was called without any variables ). So I thought perhaps I needed to add to the `useMutation` part the variables declaration which I tried in several ways. 

I also tried to circumvent the first issue by renaming the destructured object as [suggested here](https://community.apollographql.com/t/how-to-write-query-and-mutation-in-same-componnet/2159/2) with { data: queryData, ... }


I feel like I am still missing something rather fundamental here and not sure what or why.